core/connection.go

```diff
@@ -0,0 +1,134 @@
+package connection
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+)
+
+/*
+This retry library was inspired by and uses Cenk Alti (https://github.com/cenkalti) backoff library (https://github.com/cenkalti/backoff).
+We would like to thank him for his great work.
+*/
+
+type PermanentError struct {
+	Inner error
+}
+
+func (e PermanentError) Error() string { return e.Inner.Error() }
+func (e PermanentError) Unwrap() error {
+	return e.Inner
+}
+func (e PermanentError) Is(err error) bool {
+	_, ok := err.(PermanentError)
+	return ok
+}
+
+type TransientError struct {
+	Inner error
+}
+
+func (e TransientError) Error() string { return e.Inner.Error() }
+func (e TransientError) Unwrap() error {
+	return e.Inner
+}
+func (e TransientError) Is(err error) bool {
+	_, ok := err.(TransientError)
+	return ok
+}
+
+const MinDelay = 1000
+const RetryFactor = 2
+const NumRetries = 3
+const MaxInterval = 60000
+
+// Same as Retry only that the functionToRetry can return a value upon correct execution
+func RetryWithData[T any](functionToRetry func() (T, error), minDelay uint64, factor float64, maxTries uint64, maxInterval uint64) (T, error) {
```

Comment:
i dont think `retryWithData` is a good variation of the name `retry` that returns data.
`retryReturingData` or similar is better IMO

Reply:
This comes from the underlying backoff library ref: https://github.com/cenkalti/backoff/blob/v4/retry.go#L42